### ISSUE TITLE
JSON field

### ISSUE THREAD

**Issue by streamich** [Microsoft/TypeScript#27930]
## Search Terms

- JSON

## Suggestion

Type annotation for JSON in a string.

## Use Cases

Let's say you have a string which contains valid JSON object, like so:

```js
const json = '{"hello": "world"}';
```

How can you type annotate the `json` variable, currently you can mark it as `string`:

```ts
const json: string = '{"hello": "world"}';
```

Instead there could be some TypeScript language feature that helps annotate JSON in a string more precisely, for example:

```ts
const json: JSON<{hello: string}> = '{"hello": "world"}';
```

## Examples

Specify that string contains valid JSON.

```ts
let json: JSON<any>;
```

Add typings to a HTTP response body.

```ts
let responseBody: JSON<{ping: 'pong'}>;
```

Add type safety to `JSON.parse()` method.

```ts
let responseBody: JSON<{ping: 'pong'}>;
let {ping} = JSON.parse(responseBody);
```

## Checklist

My suggestion meets these guidelines:
* [x] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. new expression-level syntax)

**Comment by weswigham** [comment]
It seems like you want, generally, refinements on string types. In the vein of #6579 (for specifically regex refinements) or #4895 (for arbitrary nominal refinements).

**Comment by streamich** [comment]
@weswigham refinements on `string` type, yes, but this proposal deals specifically with JSON.

**Comment by RyanCavanaugh** [comment]
What are the use cases for writing JSON strings in code instead of writing them as parsed literals?

**Comment by streamich** [comment]
@RyanCavanaugh I have plenty of mock data for tests as JSON in strings, when you receive response from and API it could be JSON in a string, when you read from a file it could be `.json`. I'm sure there a re plenty more examples.

**Comment by RyanCavanaugh** [comment]
What I mean is, if you're writing the code, why are you writing them in the error-prone `"{ 'x': 'y'}"` form instead of the easier `{ x: 'y' }` form? I

**Comment by streamich** [comment]
@RyanCavanaugh I am not, but sometimes you receive your data in that form and you have to deal with it. For example, here is a typical AWS SQS message example:

```json
{
  "Messages": [
    {
      "Body": "{\n  \"Message\" : \"{\\\"assetId\\\":14,\\\"status\\\":\\\"Uploading\\\",\\\"updatedAt\\\":\\\"2018-10-16T08:47:43.538Z\\\"}\",\n }"
    }
  ]
}
```

*(I have removed some fields for brevity. Also, I hope all the escapings are correct. :) )*

The above is basically dobly-serialized JSON in `Messages[0].Body` field. I have no control of this format, but I would like to type annotate it somehow. For example it could be done like so:

```ts
interface Response {
  Messages: ({
    Body: JSON {
      Message: JSON {
        assetId: number;
        status: string;
        updatedAt: string;
      }
    }
  })[];
}
```

**Comment by RyanCavanaugh** [comment]
> sometimes you receive your data in that form 

Makes sense - but in that case, we can't really do any valuable typechecking of that JSON at compile-time. Or are you saying you're copying the JSON responses into your test files? Just trying to understand

**Comment by streamich** [comment]
> ... we can't really do any valuable typechecking of that JSON at compile-time.

Sure, but code can be annotated at dev time so developer can get all the code completion and error messages that are obvious from static code analysis. For example:

```ts
JSON.parse(JSON.parse(JSON.parse(message).Body).Message).assetId; // OK
JSON.parse(JSON.parse(JSON.parse(message).Body).Message).oops; // Error: ...
```

> Or are you saying you're copying the JSON responses into your test files?

Yes.

**Comment by weswigham** [comment]
SO you're saying it's be useful coupled with a `JSON.parse` overload along the lines of
```ts
declare function parse<T>(string: JSON T): T;
```

**Comment by streamich** [comment]
@weswigham exactly!

**Comment by weswigham** [comment]
Along the lines of what people have said in #4895, you can get pretty close with branded strings today:
```ts
type JSONString<T> = string & { " __JSONBrand": T };
function parse<T>(str: JSONString<T>): T { return JSON.parse(str as string) as any; };
let responseBody = '{"ping": "pong"}' as JSONString<{ping: 'pong'}>;
parse(responseBody).ping; // OK
```
there's no automatic creation of them and no automatic validation that your string actually meets the constraint you want the type to imply, but you _can_ flow the type around, at least.

**Comment by streamich** [comment]
```ts
declare function serialize<T>(obj: jsontype T): T;
```

**Comment by streamich** [comment]
@weswigham How would you annotate `JSON.stringify` method using branded strings?

**Comment by streamich** [comment]
OK, if anyone is interested, here is what I did:

```ts
type JSON<T> = string & {__JSON__: T};
type JSONParse = <T>(str: JSON<T>) => T;
type JSONStringify = <T>(obj: T) => JSON<T>;

const JSONSTR: {
  parse: JSONParse;
  stringify: JSONStringify;
} = {
  parse: (...args) => JSON.parse(...args),
  stringify: (...args) => JSON.stringify(...args) as any,
};
```

Autocompletion works:

![image](https://user-images.githubusercontent.com/9773803/47050256-fa0d9700-d19f-11e8-9a8d-26a4ac5e9c9d.png)

**Comment by streamich** [comment]
Autocompletion for above mentioned example works:

![image](https://user-images.githubusercontent.com/9773803/47050554-1e1da800-d1a1-11e8-9685-5cf037cbba80.png)

**Comment by streamich** [comment]
BTW, create this tiny NPM package if anyone needs branded JSON strings:

https://github.com/streamich/ts-brand-json

**Comment by NotWearingPants** [comment]
It is faster to use `JSON.parse` of a string literal than to use a JSON object literal:
https://v8.dev/blog/cost-of-javascript-2019#json

So this feature is now a bit more useful (although it is better if the compiler will generate the `JSON.parse` itself when it sees a JSON literal)
